kotest-framework/kotest-framework-api/src/commonMain/kotlin/io/kotest/core/spec/lifecycle.kt

```diff
@@ -69,11 +69,24 @@ suspend fun TestCase.invokeAllBeforeTestCallbacks(): Try<TestCase> =
       Try.Failure(it)
    }, { listeners ->
       Try {
+         var currentStatus: TestStatus? = null
+         var currentException: Error? = null
          listeners.forEach {
-            if (type == TestType.Container) it.beforeContainer(this)
-            if (type == TestType.Test) it.beforeEach(this)
-            it.beforeAny(this)
-            it.beforeTest(this)
+            try {
+               if (type == TestType.Container) it.beforeContainer(this)
+               if (type == TestType.Test) it.beforeEach(this)
+               it.beforeAny(this)
+               it.beforeTest(this)
+            } catch (e: Error) {
+               if (currentStatus == null || currentStatus!! != TestStatus.Failure) {
+                  currentStatus = TestStatus.Failure
+                  currentException = e
+               }
+            }
+         }
+
+         currentException?.let { it ->
+            throw Error(it)
          }
 
```

Comment:
I think a cleaner way to handle this might look something like this.
```kotlin
listeners.map {
         Try {
            if (type == TestType.Container) it.beforeContainer(this)
            if (type == TestType.Test) it.beforeEach(this)
            it.beforeAny(this)
            it.beforeTest(this)
            this
         }
      }.find { it.isFailure() } ?: Try { this }
```

1) Its avoiding putting try-catch block and using existing Try.
2) Its avoiding mutation.

Reply:
with this decision we have problems:
[ERROR] C:\projects\totest\io\kotest\core\spec\lifecycle.kt:[71,9] Type mismatch: inferred type is Try<TestCase>? but Try.Failure? was expected
[ERROR] C:\projects\kotest\io\kotest\core\spec\lifecycle.kt:[79,38] Type mismatch: inferred type is Try<TestCase> but Try.Failure was expected

Reply:
Interesting i don't see that error in my local i made that change and pushed to my fork. Here is the github action link. All test passes, the only failure(expected failure) is for publishing the artifacts to repository.

https://github.com/ashishkujoy/kotest/runs/1180868149
https://github.com/ashishkujoy/kotest/commit/7f75d89742c06f016864a4400af486fcb28920c4